### PR TITLE
Add `@repeat` as shorthand `for` and `while` loops

### PR DESCRIPTION
**Pull request by IanButterworth**
I'd be surprised if this hadn't been suggested before but I couldn't find an issue/PR.

Just introduces the form
```julia
julia> @repeat 2 println("hello")
hello
hello

julia> @repeat 10 move_robot()

julia> @repeat Dates.now() < Dates.DateTime(2029) move_robot()

```
in place of
```julia
julia> for _ in 1:2
           println("hello")
       end
hello
hello

julia> for _ in 1:10
           move_robot()
       end

julia> while Dates.now() < Dates.DateTime(2029)
           move_robot()
       end
```

or
```julia
julia> for _ in 1:10 println("hello"); end
hello
hello

julia> for _ in 1:10 move_robot(); end

julia> while Dates.now() < Dates.DateTime(2029) move_robot(); end

```

I wondered if `@loop` would be a better name?

### CODE REVIEW

**Comment by DilumAluthge** [comment]
How does this compare to and contrast with `@nexprs`?

**Comment by KristofferC** [comment]
> Seems similar to `Base.Cartesian.@nexprs`?

`@nexprs` duplicate the expression `n` times:

```
julia> @macroexpand Base.Cartesian.@nexprs 4 i -> y[i] = A[i+j]
quote
    y[1] = A[1 + j]
    y[2] = A[2 + j]
    y[3] = A[3 + j]
    y[4] = A[4 + j]
end
```

This creates a loop.

**Comment by IanButterworth** [comment]
Perhaps `@loop` is a better name, to make that clearer

**Comment by simeonschaub** [comment]
Is this pattern really that common? I am not sure we really need this in Base since the loop is not hard to write. If someone really uses this that often, it is easy to just define this macro yourself.

File: base/essentials.jl [hunk 1]
```diff
@@ -590,6 +590,33 @@ macro goto(name::Symbol)
...
 
+"""
+    @repeat n call
+    @repeat bool_expr call
+
+Repeat `call` `n` times and discard the output.
+If an expression that returns a boolean is given as a first argument, repeat while true and disregard output.
+"""
+macro repeat(terms, ex)
+    if isa(terms, Integer)
```
Comment (simeonschaub) [inline base/essentials.jl:601]: It seems confusing to me to have a different behavior depending on whether the first argument is a literal integer here. I would expect
```julia
n = 2
@repeat n println("foo")
```
to work differently. Also note that `Bool <: Integer`, so `@repeat true foo()` will only run once and not be an infinite loop.

**Comment by IanButterworth** [comment]
One use case I've wanted something shorter for is testing hardware interactions in the repl. 

Another more minor is profiling when needing to sample something multiple times to get enough samples

```
julia> @profile @loop 100 foo()
```

File: base/essentials.jl (hunk 1 above)
Comment (IanButterworth) [reply base/essentials.jl]: Ah yes. Given the expr could return anything, I've made it a requirement that the first expr returns a Bool and assumed a while loop is wanted.
But it is unfortunately odd UX that the macro can't infer what to do here properly
```julia
julia> @repeat 3 + 1 println("hello")
ERROR: TypeError: non-boolean (Int64) used in boolean context
Stacktrace:
 [1] top-level scope
   @ ./essentials.jl:613
```

I've fixed `@repeat true foo()`, thanks for that pointer, and added `@repeat foo()`

**Comment by fingolfin** [comment]
I feel this is one of those cases were adding these shortcuts makes the lange more complex for learners -- it's yet another bit of syntax one has to learn to understand code. Frankly, I wish Julia had a few less "clever tricks" one has to learn about in order to understand average code (and instead better documentation, with more cross references and examples).

**Comment by IanButterworth** [comment]
Ok yeah this doesn't seem like a good/popular idea for Base. 

One for startup.jl instead

**Comment by StefanKarpinski** [comment]
For what it's worth, I do kind of like this.

**Comment by KristofferC** [comment]
`foreach` feels a bit similar to the integer version.

**Comment by IanButterworth** [comment]
If anyone is interested in this, I just pushed some improvements to the branch.
```julia
julia> @repeat println("hello")
hello
hello
^ChelloERROR: InterruptException:
...

julia> @repeat 3 println("hello")
hello
hello
hello

julia> @repeat 2 + 1 println("hello")
hello
hello
hello

julia> @repeat rand() > 0.5 println("hello")
hello
hello
```

Feel free to reopen

**Pull request closed without merging**